### PR TITLE
In-memory vector db

### PR DESCRIPTION
**Pull request by andrewjschuang**
### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 6c84662</samp>

This pull request enhances the code generation module for GitHub issue components by using a `prompt` parameter that filters the content of the URLs based on the similarity to the desired component. It also adds a new module `helpers.embeddings_similarity_search` that uses OpenAI Codex to get the embeddings and similarity of texts. It modifies the files `generate_for_github_issue.py`, `generate.py`, `main.py`, and adds the file `embeddings_similarity_search.py`.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 6c84662</samp>

> _To generate code for a component_
> _We use a prompt as an argument_
> _We filter the URLs_
> _With `get_relevant_docs`_
> _And use Codex for the content_

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 6c84662</samp>

*  Add a new module `helpers.embeddings_similarity_search` that provides functions to filter the content of the URLs based on the similarity to the prompt using OpenAI Codex embeddings and cosine similarity ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-fd6e3e827a9f41f0d4c92f670a15bfcdeb56583c2fb02637d79a321331e13d9cR1-R62))
*  Add a new parameter `prompt` to the `main` function in `generate.py` that receives the text that describes the desired component to be generated ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-57c6ea428b4134afc63a249ce4461df96eb6f49a0e7083c5badc9be262291a87L22-R23))
*  Pass the `prompt` parameter to the `parse_urls` function in `generate.py` that uses the `get_relevant_docs` function from `helpers.embeddings_similarity_search` to filter the content of the URLs ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-57c6ea428b4134afc63a249ce4461df96eb6f49a0e7083c5badc9be262291a87L31-R32), [link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-57c6ea428b4134afc63a249ce4461df96eb6f49a0e7083c5badc9be262291a87L87-R88), [link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-57c6ea428b4134afc63a249ce4461df96eb6f49a0e7083c5badc9be262291a87L97-R102))
*  Extract the `prompt` variable from the `extract_prompts` function in `generate_for_github_issue.py` and pass it to the `generate_app_file_prompt` function that uses OpenAI Codex to generate the app file instructions ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-69bedac25cab56c4653721b18b35683ac9f1958fcf91f090470a07db747ee17bL110-R111), [link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-69bedac25cab56c4653721b18b35683ac9f1958fcf91f090470a07db747ee17bR117))
*  Pass the `prompt` parameter to the `main` function in `generate_for_github_issue.py` that calls the `main` function in `generate.py` with the filtered URLs and the app file instructions ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-69bedac25cab56c4653721b18b35683ac9f1958fcf91f090470a07db747ee17bL167-R168), [link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-69bedac25cab56c4653721b18b35683ac9f1958fcf91f090470a07db747ee17bL187-R190))
*  Import the `get_relevant_docs` function from `helpers.embeddings_similarity_search` in `generate.py` ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-57c6ea428b4134afc63a249ce4461df96eb6f49a0e7083c5badc9be262291a87R7))
*  Remove the `--no-verify` flag from the `git push` command in `generate_for_github_issue.py` as the code is assumed to be valid and formatted by OpenAI Codex ([link](https://github.com/PipedreamHQ/pipedream/pull/8993/files?diff=unified&w=0#diff-69bedac25cab56c4653721b18b35683ac9f1958fcf91f090470a07db747ee17bL187-R190))

### CODE REVIEW

File: packages/component_code_gen/helpers/embeddings_similarity_search.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    if not response.ok:
+        raise Exception(
+            f"OpenAI exception: {response.status_code} - {response.text}")
+
+    embedding = response.json()["data"][0]["embedding"]
+    return np.asarray(embedding).reshape(1, -1)
+
+
+def split_document(document, chunk_size=1000):
```
Comment (dylburger) [inline packages/component_code_gen/helpers/embeddings_similarity_search.py:34]: You might try using https://python.langchain.com/docs/modules/data_connection/document_transformers/text_splitters/HTML_header_metadata or thinking about how to split the documents into logical sections vs. just chunking on size. This is fine for now and I would merge if you're seeing perf improvements, but chunking is so critical that we can iterate on it in the future and see if we see further increases.

File: packages/component_code_gen/main.py [hunk 2]
```diff
@@ -43,6 +43,6 @@
         with open(args.instructions, 'r') as f:
             instructions = f.read()
 
-        result = main(args.type, args.app, instructions, args.tries,
+        result = main(args.type, args.app, instructions, instructions, args.tries,
```
Comment (dylburger) [inline packages/component_code_gen/main.py:46]: If we're passing the same value for both the `instructions` and `prompt` args now, is it worth refactoring to only pass instructions? Or is there still a reason to keep them separate?

**Review by dylburger — APPROVED** [review]
lgtm. I believe you should merge what you have now, but see my two comments for future improvements.

File: packages/component_code_gen/main.py (hunk 2 above)
Comment (andrewjschuang) [reply packages/component_code_gen/main.py:46]: This was added to the previous way we called the script, by passing a local instructions.md file to it.

I think we can refactor to basically remove everything except for Github Issues. It's the easiest approach on how we interface with the script. Wdyt?

**Pull request merged**